src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -0,0 +1,683 @@
+			s_IL2080, s_IL2081, s_IL2082, s_IL2083, s_IL2084, s_IL2085, s_IL2086, s_IL2087, s_IL2088, s_IL2089,
+			s_IL2090, s_IL2091);
+		#endregion
+
+		public override void Initialize (AnalysisContext context)
+		{
+			context.EnableConcurrentExecution ();
+			context.ConfigureGeneratedCodeAnalysis (GeneratedCodeAnalysisFlags.ReportDiagnostics);
+			context.RegisterSyntaxNodeAction (DynamicallyAccessedMembersAnalyze,
+				SyntaxKind.InvocationExpression,
+				SyntaxKind.ReturnStatement,
+				SyntaxKind.SimpleAssignmentExpression);
+		}
+
+		static void DynamicallyAccessedMembersAnalyze (SyntaxNodeAnalysisContext context)
+		{
+			var syntaxNode = context.Node;
+			if (syntaxNode.AncestorsAndSelf ().OfType<StatementSyntax> ().FirstOrDefault () is not StatementSyntax statementSyntax)
+				return;
+
+			var cfAnalysis = context.SemanticModel.AnalyzeControlFlow (statementSyntax);
+			// We don't care about unreachable statements.
+			if (cfAnalysis is null || !cfAnalysis.Succeeded || !cfAnalysis.StartPointIsReachable)
+				return;
+
+			switch (syntaxNode) {
+			case InvocationExpressionSyntax invocationExpression:
+				ProcessInvocationExpression (invocationExpression, context);
+				break;
+			case ReturnStatementSyntax returnStatement:
+				ProcessReturnStatement (returnStatement, context);
+				break;
+			case AssignmentExpressionSyntax assignmentExpression:
+				ProcessAssignmentExpression (assignmentExpression, context);
+				break;
+			}
+		}
+
+		static void ProcessInvocationExpression (InvocationExpressionSyntax invocationExpression, SyntaxNodeAnalysisContext context)
+		{
+			if (context.SemanticModel.GetSymbol (invocationExpression) is not IMethodSymbol invokedMethodSymbol)
+				return;
+
+			if (invocationExpression.Expression is MemberAccessExpressionSyntax memberAccessExpression)
+				ProcessMemberAccessExpression (memberAccessExpression, invokedMethodSymbol, context);
+
+			// For all generic parameters in the called method, check if the passed type arguments have matching annotations.
+			for (int i = 0; i < invokedMethodSymbol.TypeParameters.Length; i++) {
+				var typeParameterSymbol = invokedMethodSymbol.TypeParameters[i];
+				if (!typeParameterSymbol.TryGetDynamicallyAccessedMemberTypes (out var damtOnTypeParameter))
+					continue;
+
+				var typeArgumentSymbol = invokedMethodSymbol.TypeArguments[i];
+				typeArgumentSymbol.TryGetDynamicallyAccessedMemberTypes (out var damtOnTypeArgument);
+				if (!SourceHasMatchingAnnotations (damtOnTypeArgument, damtOnTypeParameter, out var missingAnnotationsOnTypeArgument)) {
+					ReportUnmatchedAnnotations (invocationExpression.GetLocation (),
+						typeArgumentSymbol, typeParameterSymbol, missingAnnotationsOnTypeArgument, context);
+					continue;
+				}
+			}
+
+			// For all parameters in the called method, check if the passed arguments have matching annotations.
+			for (int i = 0; i < invokedMethodSymbol.Parameters.Length; i++) {
```

Comment:
I think what you want to use here is an `IInvocationOperation`, because then you can loop through the `IArgumentOperations`. Otherwise you'll be stuck trying to match up arguments with parameters, which is just insanely difficult given things like named arguments, optional parameters, `params` arguments, etc.